src/compiler/types.ts

```diff
@@ -1271,16 +1250,83 @@ namespace ts {
         | SyntaxKind.CommaToken
         ;
 
-    export type BinaryOperatorToken = Token<BinaryOperator>;
-
-    export interface BinaryExpression extends Expression, Declaration  {
+    export type BinaryOperatorToken =

```

Comment:
Don't see why we need unions for the tokens; it would be easier to write `Token<BinaryOperator>`.

Reply:
Token<BinaryOperator> is not assignable to the union of all tokens of the binary operator kinds because we do not multiply out/canonicalize unions prior to comparison checking. As such, it must be written this way to work around that.

Reply:
I actually tried this out myself and it compiled successfully.